src/System.Linq.Parallel/tests/QueryOperators/SelectSelectManyTests.cs

```diff
@@ -0,0 +1,641 @@
+            int seen = 0;
+            Assert.All(query.Select((x, index) => KeyValuePair.Create(x, index)).ToList(),
+                p =>
+                {
+                    Assert.Equal(seen++, p.Key);
+                    Assert.Equal(p.Key, p.Value);
+                });
+            Assert.Equal(count, seen);
+        }
+
+        [Theory]
+        [OuterLoop]
+        [MemberData("Ranges", (object)(new int[] { 1024, 1024 * 4 }), MemberType = typeof(Sources))]
+        public static void Select_Indexed_NotPipelined_Longrunning(Labelled<ParallelQuery<int>> labelled, int count)
+        {
+            Select_Indexed_NotPipelined(labelled, count);
+        }
+
+        [Fact]
+        public static void Select_ArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>(() => ((ParallelQuery<bool>)null).Select(x => x));
+            Assert.Throws<ArgumentNullException>(() => ((ParallelQuery<bool>)null).Select((x, index) => x));
+            Assert.Throws<ArgumentNullException>(() => ParallelEnumerable.Empty<bool>().Select((Func<bool, bool>)null));
+            Assert.Throws<ArgumentNullException>(() => ParallelEnumerable.Empty<bool>().Select((Func<bool, int, bool>)null));
+        }
+
+        //
+        // SelectMany
+        //
+
+        public static IEnumerable<object[]> SelectManyUnorderedData(object[] counts)
+        {
+            foreach (object[] results in UnorderedSources.Ranges(counts.Cast<int>()))
+            {
+                foreach (Labelled<Func<int, int, IEnumerable<int>>> expander in Expanders())
+                {
+                    foreach (int count in new[] { 0, 1, 2, 8 })
+                    {
+                        yield return new object[] { results[0], results[1], expander, count };
+                    }
+                }
+            }
+        }
+
+        public static IEnumerable<object[]> SelectManyData(object[] counts)
+        {
+            foreach (object[] results in Sources.Ranges(counts.Cast<int>()))
+            {
+                foreach (Labelled<Func<int, int, IEnumerable<int>>> expander in Expanders())
+                {
+                    foreach (int count in new[] { 0, 1, 2, 8 })
+                    {
+                        yield return new object[] { results[0], results[1], expander, count };
+                    }
+                }
+            }
+        }
+
+        private static IEnumerable<Labelled<Func<int, int, IEnumerable<int>>>> Expanders()
+        {
+            yield return Labelled.Label("Array", (Func<int, int, IEnumerable<int>>)((start, count) => Enumerable.Range(start * count, count).ToArray()));
+            yield return Labelled.Label("Enumerable.Range", (Func<int, int, IEnumerable<int>>)((start, count) => Enumerable.Range(start * count, count)));
+            yield return Labelled.Label("ParallelEnumerable.Range", (Func<int, int, IEnumerable<int>>)((start, count) => ParallelEnumerable.Range(start * count, count)));
```

Comment:
Should we add some types besides arrays, enumerables, and parallel enumerables? e.g. add a list, a read-only collection, etc.?